how-migration-discovery-should-work.md

```diff
@@ -0,0 +1,74 @@
+# How migration discovery should work
+
+Let's look at an example task that has had multiple versions and migrations.
+
+```text
+task/source-build
+├── 0.2
+│   ├── migrations
+│   │   │── 0.2.sh
+│   │   └── 0.2.1.sh
+│   └── source-build.yaml   <- has app.kubernetes.io/version: "0.2.1"
+└── 0.3
+    ├── migrations
+    │   │── 0.3.sh
+    │   └── 0.3.1.sh
+    └── source-build.yaml   <- has app.kubernetes.io/version: "0.3.1"
+```
+
+* We build both versions of `task/source-build`.
+* When building `task/source-build/0.2`, we look at the version label and see `0.2.1`.
+  * We attach `migrations/0.2.1.sh`
+  * We tag it with `:0.2.1` and `:0.2.1-{timestamp or git_sha or something unique}`
+* Same for `task/source-build/0.3`, but with `0.3.1`
+
+Let's say we've been doing it this way for a while and have released all the versions
+(`0.2`, `0.2.1`, `0.3`, `0.3.1`) a couple times. In the registry, the situation may
+look something like this:
+
+```text
+quay.io/konflux-ci/tekton-catalog/task-source-build
+
+  "the 0.2 branch"                          "the 0.3 branch"
+
+      0.2.1-c   <- migrations/0.2.1.sh
+        |                                       0.3.1-b     <- migrations/0.3.1.sh
+        |                                         |
+        |                                       0.3.1-a     <- migrations/0.3.1.sh
+      0.2.1-b   <- migrations/0.2.1.sh            |
+        |                                         |
+      0.2.1-a   <- migrations/0.2.1.sh            |
+        |                                        0.3-b      <- migrations/0.3.sh
+        |                                         |
+        |                                        0.3-a      <- migrations/0.3.sh
+       0.2-b    <- migrations/0.2.sh
+        |
+       0.2-a    <- migrations/0.2.sh
+```
+
+*In reality, this would be a single linear timeline of tags. It's split in
+two "branches" here for easier visualization.*
```

Comment:
Out-of-order tags should be considered. Since each task version is a Component, they are built separately. The order of tags in the release image repo is not guaranteed to be a single linear timeline. Unless this is not important for migration discovery.